service/src/main/kotlin/fi/espoo/evaka/shared/data/RangeBasedSet.kt

```diff
@@ -0,0 +1,101 @@
+    fun addAll(ranges: Sequence<Range>): This = ranges.fold(this.ranges, ::add).toThis()
+
+    fun remove(range: Range): This = remove(this.ranges, range).toThis()
+    fun removeAll(vararg ranges: Range): This = ranges.fold(this.ranges, ::remove).toThis()
+    fun removeAll(ranges: Iterable<Range>): This = ranges.fold(this.ranges, ::remove).toThis()
+    fun removeAll(ranges: Sequence<Range>): This = ranges.fold(this.ranges, ::remove).toThis()
+
+    fun contains(range: Range): Boolean = this.ranges.any { it.contains(range) }
+    fun includes(date: Point): Boolean = this.ranges.any { it.includes(date) }
+
+    fun intersection(other: Iterable<Range>): This =
+        intersection(this.ranges.iterator(), other.iterator()).toThis()
+    fun intersection(other: Sequence<Range>): This =
+        intersection(this.ranges.iterator(), other.iterator()).toThis()
+
+    protected abstract fun List<Range>.toThis(): This
+    protected abstract fun range(start: Point, end: Point): Range
+
+    companion object {
+        fun <Point : Comparable<Point>, Range : BoundedRange<Point, Range>> add(
+            ranges: List<Range>,
+            range: Range
+        ): List<Range> =
+            ranges
+                .partition { it.overlaps(range) || it.adjacentTo(range) }
+                .let { (conflicts, unchanged) ->
+                    val result = unchanged.toMutableList()
+                    result.add(conflicts.fold(range) { acc, it -> acc.merge(it) })
+                    result.sortBy { it.start }
+                    result
+                }
+
+        fun <Point : Comparable<Point>, Range : BoundedRange<Point, Range>> remove(
+            ranges: List<Range>,
+            range: Range
+        ): List<Range> =
+            ranges
+                .partition { it.overlaps(range) }
+                .let { (conflicts, unchanged) ->
+                    val result = unchanged.toMutableList()
+                    for (conflict in conflicts) {
+                        result.addAll(conflict.subtract(range))
+                    }
+                    result.sortBy { it.start }
+                    result
+                }
+
+        fun <Point : Comparable<Point>, Range : BoundedRange<Point, Range>> intersection(
+            iterA: Iterator<Range>,
+            iterB: Iterator<Range>
+        ): List<Range> {
+            var a = if (iterA.hasNext()) iterA.next() else null
+            var b = if (iterB.hasNext()) iterB.next() else null
+            val ranges = mutableListOf<Range>()
+            while (a != null && b != null) {
+                a.intersection(b)?.let { ranges.add(it) }
+                if (a.end <= b.end) {
+                    a = if (iterA.hasNext()) iterA.next() else null
+                } else {
+                    b = if (iterB.hasNext()) iterB.next() else null
+                }
+            }
+            return ranges
+        }
+    }
```

Comment:
this was impossible for me to follow without explanation

Reply:
Tidied up slightly and added a reference link for inspiration